File: Level 3/여행경로/solution.js
function solution(tickets) {
  const answer = [];

  const dfs = (from, remainTickets, path) => {
    const updatedPath = [...path, from];

    if (!remainTickets.length) {
      answer.push(updatedPath);
    } else {
      remainTickets.forEach((ticket, index) => {
        if (ticket[0] !== from) return;

        const to = ticket[1];
        const updatedRemainTickets = [...remainTickets];
        updatedRemainTickets.splice(index, 1);

        dfs(to, updatedRemainTickets, updatedPath);
      });
    }
  };

  dfs("ICN", tickets, []);

  return answer.sort()[0];
}

File: Level 3/여행경로/solution.py
from collections import defaultdict


def build_graph(tickets):
    graph = defaultdict(list)

    for start, end in tickets:
        graph[start].append(end)

    for key in graph.keys():
        graph[key].sort()

    return graph


def dfs(start, path, graph, N):
    if len(path) == N:
        return path

    for index, end in enumerate(graph[start]):
        graph[start].pop(index)

        new_path = path[:] + [end]

        result = dfs(end, new_path, graph, N)

        graph[start].insert(index, end)

        if result:
            return result


def solution(tickets):
    graph = build_graph(tickets)
    N = len(tickets) + 1

    answer = dfs("ICN", ["ICN"], graph, N)

    return answer
